### PR TITLE
feat: Support JS symbolication in symbolicli

### PR DESCRIPTION
**Pull request by loewenheim**
This enables `symbolicli` to process JS events. Making this work requires a bit of restructuring; fetching/parsing the event is now nearly the first thing we do because we need to know whether we're dealing with a JS or native event. It also
* moves the `output` module into its own file since it's getting pretty chonky;
* drive-by implements `Display` for `JsModuleErrorKind` so `symbolicli` can print errors in the `pretty` format.

Question: `compact` and `pretty` output print only the first stacktrace in the JS case. For native, they print the crashing thread or else the first, if none are marked as crashing. Is this acceptable? The only alternative I can think of is just printing all stacktraces.

### CODE REVIEW

**Comment by kamilogorek** [comment]
> moves the output module into its own file since it's getting pretty chonky;

image

**Comment by kamilogorek** [comment]
> Question: compact and pretty output print only the first stacktrace in the JS case. For native, they print the crashing thread or else the first, if none are marked as crashing. Is this acceptable? The only alternative I can think of is just printing all stacktraces.

This is also the current behavior of `explain` and `/source-map-debug` endpoint, which I think is fine fo rnow.

**Comment by krystofwoldrich** [comment]
I've tried it with an RN event but got this error. RN set the platform to `node` in the event. 

```
 ❯ target/release/symbolicli --org sentry-sdks --project sentry-react-native ccba9b24c81b4adc8b46ea1c002420e8
2023-05-12T07:45:37.522284Z  INFO symbolicli: event not found in local file system
2023-05-12T07:45:37.522416Z  INFO symbolicli: trying to resolve event remotely
2023-05-12T07:45:37.522433Z  INFO symbolicli::remote: downloading event json url=https://sentry.io/api/0/projects/sentry-sdks/sentry-react-native/events/ccba9b24c81b4adc8b46ea1c002420e8/json/
Error: Failed to decode event

Caused by:
    0: error decoding response body: unknown variant `node`, expected `native` or `javascript` at line 1 column 137
    1: unknown variant `node`, expected `native` or `javascript` at line 1 column 137
```

**Comment by loewenheim** [comment]
@krystofwoldrich Should be fixed now.

**Comment by krystofwoldrich** [comment]
Thank you.

```
{
    "stacktraces": [
        {
            "frames": [
                {
                    "function": "Button.props.onPress",
                    "filename": "app:///HomeScreen.tsx",
                    "module": "HomeScreen",
                    "abs_path": "app:///HomeScreen.tsx",
                    "lineno": 75,
                    "colno": 45,
                    "pre_context": [
                        "          }}",
                        "        />",
                        "        <Button",
                        "          title=\"Capture exception\"",
                        "          onPress={() => {"
                    ],
                    "context_line": "            Sentry.captureException(new Error('Captured exception'));",
                    "post_context": [
                        "          }}",
                        "        />",
                        "        <Button",
                        "          title=\"Uncaught Thrown Error\"",
                        "          onPress={() => {"
                    ],
                    "in_app": true
                },
                {
                    "function": "_performTransitionSideEffects",
                    "filename": "app:///Pressability.js",
                    "module": "Pressability",
                    "abs_path": "app:///Pressability.js",
                    "colno": 17,
                    "in_app": false
                },
                {
                    "function": "_receiveSignal",
                    "filename": "app:///Pressability.js",
                    "module": "Pressability",
                    "abs_path": "app:///Pressability.js",
                    "colno": 40,
                    "in_app": false
                },
                {
                    "function": "responderEventHandlers.onResponderRelease",
                    "filename": "app:///Pressability.js",
                    "module": "Pressability",
                    "abs_path": "app:///Pressability.js",
                    "colno": 27,
                    "in_app": false
                },
                {
                    "function": "invokeGuardedCallbackProd",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 14,
                    "in_app": false
                },
                {
                    "function": "invokeGuardedCallback",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 35,
                    "in_app": false
                },
                {
                    "function": "invokeGuardedCallbackAndCatchFirstError",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 29,
                    "in_app": false
                },
                {
                    "function": "executeDispatch",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 41,
                    "in_app": false
                },
                {
                    "function": "executeDispatchesInOrder",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 19,
                    "in_app": false
                },
                {
                    "function": "executeDispatchesAndRelease",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 28,
                    "in_app": false
                },
                {
                    "function": "executeDispatchesAndReleaseTopLevel",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 36,
                    "in_app": false
                },
                {
                    "function": "forEachAccumulated",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 15,
                    "in_app": false
                },
                {
                    "function": "runEventsInBatch",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 20,
                    "in_app": false
                },
                {
                    "function": "runExtractedPluginEventsInBatch",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 18,
                    "in_app": false
                },
                {
                    "function": "batchedUpdates$argument_0",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 35,
                    "in_app": false
                },
                {
                    "function": "batchedUpdates$1",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 13,
                    "in_app": false
                },
                {
                    "function": "batchedUpdates",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 29,
                    "in_app": false
                },
                {
                    "function": "dispatchEvent",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 16,
                    "in_app": false
                }
            ]
        }
    ],
    "raw_stacktraces": [
        {
            "frames": [
                {
                    "function": "Button.props.onPress",
                    "filename": "app:///HomeScreen.tsx",
                    "module": "HomeScreen",
                    "abs_path": "app:///HomeScreen.tsx",
                    "lineno": 75,
                    "colno": 45,
                    "pre_context": [
                        "          }}",
                        "        />",
                        "        <Button",
                        "          title=\"Capture exception\"",
                        "          onPress={() => {"
                    ],
                    "context_line": "            Sentry.captureException(new Error('Captured exception'));",
                    "post_context": [
                        "          }}",
                        "        />",
                        "        <Button",
                        "          title=\"Uncaught Thrown Error\"",
                        "          onPress={() => {"
                    ],
                    "in_app": true
                },
                {
                    "function": "_performTransitionSideEffects",
                    "filename": "app:///Pressability.js",
                    "module": "Pressability",
                    "abs_path": "app:///Pressability.js",
                    "colno": 17,
                    "in_app": false
                },
                {
                    "function": "_receiveSignal",
                    "filename": "app:///Pressability.js",
                    "module": "Pressability",
                    "abs_path": "app:///Pressability.js",
                    "colno": 40,
                    "in_app": false
                },
                {
                    "function": "responderEventHandlers.onResponderRelease",
                    "filename": "app:///Pressability.js",
                    "module": "Pressability",
                    "abs_path": "app:///Pressability.js",
                    "colno": 27,
                    "in_app": false
                },
                {
                    "function": "invokeGuardedCallbackProd",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 14,
                    "in_app": false
                },
                {
                    "function": "invokeGuardedCallback",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 35,
                    "in_app": false
                },
                {
                    "function": "invokeGuardedCallbackAndCatchFirstError",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 29,
                    "in_app": false
                },
                {
                    "function": "executeDispatch",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 41,
                    "in_app": false
                },
                {
                    "function": "executeDispatchesInOrder",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 19,
                    "in_app": false
                },
                {
                    "function": "executeDispatchesAndRelease",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 28,
                    "in_app": false
                },
                {
                    "function": "executeDispatchesAndReleaseTopLevel",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 36,
                    "in_app": false
                },
                {
                    "function": "forEachAccumulated",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 15,
                    "in_app": false
                },
                {
                    "function": "runEventsInBatch",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 20,
                    "in_app": false
                },
                {
                    "function": "runExtractedPluginEventsInBatch",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 18,
                    "in_app": false
                },
                {
                    "function": "batchedUpdates$argument_0",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 35,
                    "in_app": false
                },
                {
                    "function": "batchedUpdates$1",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 13,
                    "in_app": false
                },
                {
                    "function": "batchedUpdates",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 29,
                    "in_app": false
                },
                {
                    "function": "dispatchEvent",
                    "filename": "app:///ReactFabric-dev.js",
                    "module": "ReactFabric-dev",
                    "abs_path": "app:///ReactFabric-dev.js",
                    "colno": 16,
                    "in_app": false
                }
            ]
        }
    ],
    "errors": [
        {
            "abs_path": "app:///HomeScreen.tsx",
            "type": "missing_source"
        },
        {
            "abs_path": "app:///Pressability.js",
            "type": "invalid_location",
            "line": null,
            "col": 17
        },
        {
            "abs_path": "app:///Pressability.js",
            "type": "invalid_location",
            "line": null,
            "col": 27
        },
        {
            "abs_path": "app:///Pressability.js",
            "type": "invalid_location",
            "line": null,
            "col": 40
        },
        {
            "abs_path": "app:///ReactFabric-dev.js",
            "type": "invalid_location",
            "line": null,
            "col": 13
        },
        {
            "abs_path": "app:///ReactFabric-dev.js",
            "type": "invalid_location",
            "line": null,
            "col": 14
        },
        {
            "abs_path": "app:///ReactFabric-dev.js",
            "type": "invalid_location",
            "line": null,
            "col": 15
        },
        {
            "abs_path": "app:///ReactFabric-dev.js",
            "type": "invalid_location",
            "line": null,
            "col": 16
        },
        {
            "abs_path": "app:///ReactFabric-dev.js",
            "type": "invalid_location",
            "line": null,
            "col": 18
        },
        {
            "abs_path": "app:///ReactFabric-dev.js",
            "type": "invalid_location",
            "line": null,
            "col": 19
        },
        {
            "abs_path": "app:///ReactFabric-dev.js",
            "type": "invalid_location",
            "line": null,
            "col": 20
        },
        {
            "abs_path": "app:///ReactFabric-dev.js",
            "type": "invalid_location",
            "line": null,
            "col": 28
        },
        {
            "abs_path": "app:///ReactFabric-dev.js",
            "type": "invalid_location",
            "line": null,
            "col": 29
        },
        {
            "abs_path": "app:///ReactFabric-dev.js",
            "type": "invalid_location",
            "line": null,
            "col": 35
        },
        {
            "abs_path": "app:///ReactFabric-dev.js",
            "type": "invalid_location",
            "line": null,
            "col": 36
        },
        {
            "abs_path": "app:///ReactFabric-dev.js",
            "type": "invalid_location",
            "line": null,
            "col": 41
        }
    ]
}
```

File: crates/symbolicli/src/output.rs [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+                row.add_cell(cell!(module_file));
+                row.add_cell(cell!(r->format!("+{module_offset:#x}")));
+            }
+            None => row.add_cell(cell!("").with_hspan(2)),
+        }
+
+        match func {
+            Some((func, func_offset)) => {
+                row.add_cell(cell!(func));
+                row.add_cell(cell!(r->format!(" + {func_offset:#x}")));
```
Comment (Swatinem) [inline crates/symbolicli/src/output.rs:153]:
```suggestion
                row.add_cell(cell!(func));
                row.add_cell(cell!(r->format!("+{func_offset:#x}")));
```
To match the above?

File: crates/symbolicli/src/output.rs [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+
+        if let Some((module_file, module_offset)) = module {
+            table.add_row(row![
+                r->"  Module:",
+                format!("{module_file} +{module_offset:#0x}")
+            ]);
+        }
+
+        if let Some((func, func_offset)) = func {
+            table.add_row(row![r->"  Function:", format!("{func} + {func_offset:#x}")]);
```
Comment (Swatinem) [inline crates/symbolicli/src/output.rs:200]:
```suggestion
            table.add_row(row![r->"  Function:", format!("{func} +{func_offset:#x}")]);
```

File: crates/symbolicli/src/main.rs [hunk 3]
```diff
@@ -694,12 +626,19 @@ mod event {
...
+        module: Option<String>,
+
         source_link: Option<String>,
 
         in_app: Option<bool>,
 
         #[serde(default)]
         trust: FrameTrust,
+
+        token_name: Option<String>,
```
Comment (Swatinem) [inline crates/symbolicli/src/main.rs:638]: That is not a "public" field though, its only internal. I would say to just make them `None` in `to_js_frame`.

**Review by Swatinem — APPROVED** [review]

**Pull request merged**